### PR TITLE
Clone PerMessageDeflateExtension values correctly

### LINKED ISSUE #1400: PerMessageDeflateExtension#setDeflater()/#setInflater() is overwritten in case of no_context_takeover
**Describe the bug**
In the PerMessageDeflateExtension class, the setDeflater() and setInflater() methods set the internal deflater and inflater objects. However, in case of client|serverNoClientTakeover being true (the latter is the default), the set objects will be overwritten internally by their hardcoded defaults. As this happens during data exchange, there is no obvious hook by which the application could repeat the setDeflater() and setInflater() calls. 

Relevant excerpts from PerMessageDeflateExtension.java:

  private boolean serverNoContextTakeover = true;
  private boolean clientNoContextTakeover = false;

  private Inflater inflater = new Inflater(true);
  private Deflater deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);

  public void setInflater(Inflater inflater) {
    this.inflater = inflater;
  }

  public void setDeflater(Deflater deflater) {
    this.deflater = deflater;
  }

  @Override
  public void decodeFrame(Framedata inputFrame) throws InvalidDataException {
...
        // If context takeover is disabled, inflater can be reset.
        if (clientNoContextTakeover) {
          inflater = new Inflater(true);
        }
...

  @Override
  public void encodeFrame(Framedata inputFrame) {
...
      if (serverNoContextTakeover) {
        deflater.end();
        deflater = new Deflater(Deflater.DEFAULT_COMPRESSION, true);
      }
...

**Expected behavior**
The Inflater and Deflater objects set once should be retained throughout the lifetime of a class instance. I'm not sure why the current implementation creates new objects instead of calling inflater.reset() and deflater.reset(), respectively. If the reset calls achieve the same result as recreating the objects, it is a simple two-line fix. If, however, new object instances are indeed required, the API would have to be changed to pass builder objects instead of object instances.

### LINKED ISSUE #1437: Question: How can the compression threshold be set for the PerMessageDeflateExtension in a Deflate Client?
**Describe what you would like to know or do**
I'd like to be able to set the compression threshold to a lower value than the 1024 byte value in the PerMessageDeflateExtension.

**Describe the solution you'd considered**
I've taken a copy of the 1.5.7 source code and changed the default threshold which does work but this is not ideal.

**Additional context**
I'm sending messages which are a little smaller than 1024 bytes that I'd like to compress.

### PR DESCRIPTION
**Pull request by marci4**
## Related Issue

Fixes #1437
Fixes #1400

## Motivation and Context

Issue

## How Has This Been Tested?

Additional tests

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

**Review by PhilipRoman — APPROVED** [review]

File: src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java [hunk 1]
```diff
@@ -330,7 +330,11 @@ public String getProvidedExtensionAsServer() {
 
   @Override
   public IExtension copyInstance() {
-    return new PerMessageDeflateExtension();
+    PerMessageDeflateExtension clone = new PerMessageDeflateExtension();
+    clone.setThreshold(this.getThreshold());
```
Comment (robert-s-ubi) [inline src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:334]: Please insert:

+    clone.setDeflater(this.getDeflater());
+    clone.setInflater(this.getInflater());

and replace the new Inflater() calls in lines 169 and 177 with inflater.reset(), and the deflater.end() & new Deflater() calls in lines 247/248 with deflater.reset().

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:334]: We are not going to reuse the deflater/inflater for multiple clients

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:334]: Using reset however should be done, I agree

Comment (robert-s-ubi) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: > We are not going to reuse the deflater/inflater for multiple clients

Could you elaborate? If the user of this library wants to hand in a different inflater/deflater, why should that not be used?

Specific use case: I want to use BEST_COMPRESSION, not DEFAULT_COMPRESSION, to get the most out of compression (and both my servers and clients are powerful enough to put up that extra effort). That's why I hand in a new Deflater(Deflater.BEST_COMPRESSION, true) instance. I'd like that to work.

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: The problem is that we create a new instance of a Draft (and therefore the provided extension with the custom deflater/inflater) for every new client connecting to the server.
https://github.com/TooTallNate/Java-WebSocket/blob/master/src/main/java/org/java_websocket/WebSocketImpl.java#L272

If we simply use the single instance provided by the user, we will get in trouble for the server implementation

Comment (robert-s-ubi) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: > If we simply use the single instance provided by the user, we will get in trouble for the server implementation

Ah, thanks, I see. I was just lucky that I was only using 1 server and client for my WebSocket compression tests.

Indeed, then an API change is needed to set Inflater and Deflater factories. Or, if we think that this flexibility is not needed, a setCompressionLevel(int level); API would do for me. Then this class would always instantiate the Inflater and Deflater objects, always with nowrap=true.

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: Looking more into the documentation, nowrap will always be true in our case

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: > > If we simply use the single instance provided by the user, we will get in trouble for the server implementation
> 
> Ah, thanks, I see. I was just lucky that I was only using 1 server and client for my WebSocket compression tests.
> 
> Indeed, then an API change is needed to set Inflater and Deflater factories. Or, if we think that this flexibility is not needed, a setCompressionLevel(int level); API would do for me. Then this class would always instantiate the Inflater and Deflater objects, always with nowrap=true.

Yeah I think so too. We only need the compression level, nowrap should be true.

@PhilipRoman you got any objections?

Comment (marci4) [reply src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:333]: My bad, there is no getter for the level/strategy.
setCompressionLevel() is the way to go

**Comment by PhilipRoman** [comment]
Looks good, although personally I would add some calls to setDeflaterLevel in the functional tests like testDecodeFrame.

File: src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java [hunk 2]
```diff
@@ -53,23 +53,28 @@ public class PerMessageDeflateExtension extends CompressionExtension {
...
-  public void setDeflater(Deflater deflater) {
-    this.deflater = deflater;
+  /**
+   * Set the compression level used for the compressor.
+   * @param level the compression level (0-9)
+   */
+  public void setDeflaterLevel(int level) {
+    this.deflater.setLevel(level);
+    this.deflaterLevel = level;
+    //If the compression level is changed, the next invocation of deflate will compress the input available so far with the old level (and may be flushed); the new level will take effect only after that invocation.
```
Comment (robert-s-ubi) [inline src/main/java/org/java_websocket/extensions/permessage_deflate/PerMessageDeflateExtension.java:76]: Unfortunately, this does not work in Java 8:

While this matches the Java 8 Deflater Javadoc, the implementation does not match the documentation:
https://github.com/openjdk/jdk8u/blob/master/jdk/src/share/classes/java/util/zip/Deflater.java

The set level is never picked up as far as I can tell. The "setParams" boolean is never evaluated. It is in Java 11, though, so it works there:
https://github.com/openjdk/jdk11u/blob/master/src/java.base/share/classes/java/util/zip/Deflater.java

So, for Java 8 compatibility, the Deflater instance has to be created with the configured compression level.

How about not having setDeflaterLevel at all, but rather making it an optional constructor argument:

private final int compressionLevel;
private final Inflater inflater;
private final Deflater deflater;

public PerMessageDeflateExtension() {
  this(Deflater.DEFAULT_COMPRESSION);
}

public PerMessageDeflateExtension(int compressionLevel) {
  this.compressionLevel = compressionLevel;
  this.inflater = new Inflater(true);
  this.deflater = new Deflater(compressionLevel, true);
}

public int getCompressionLevel() {
  return compressionLevel;
}

  @Override
  public IExtension copyInstance() {
    PerMessageDeflateExtension clone = new PerMessageDeflateExtension(this.getCompressionLevel());
     clone.setThreshold(this.getThreshold());
    clone.setClientNoContextTakeover(this.isClientNoContextTakeover());
     clone.setServerNoContextTakeover(this.isServerNoContextTakeover());
     return clone;
   }

**Review by robert-s-ubi — CHANGES_REQUESTED** [review]
Unfortunately, changes are needed for Java 8 compatibility.

**Review by robert-s-ubi — APPROVED** [review]
Looks good to me. Thanks for your patience!

**Pull request merged**